Pass modules to PythonExpression

RobertBlakeAnderson opened the pull request:
Added ability to pass Python modules to the PythonExpression substitution.
Allows eval of more expressions.

For example, my motivating use case was:
`robot_desc = PythonExpression(["process_file('", filepath, "').toprettyxml(indent='  ')"], xacro.__dict__)`

methylDragon commented:
Could you add tests?
@RobertBlakeAnderson

RobertBlakeAnderson commented:
Sure thing! I'll get to it shortly.

RobertBlakeAnderson commented:
Done. I also fixed a couple tiny issues that I found while creating and running the tests.

methylDragon commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17827)](http://ci.ros2.org/job/ci_linux/17827/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=12394)](http://ci.ros2.org/job/ci_linux-aarch64/12394/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=18496)](http://ci.ros2.org/job/ci_windows/18496/)

methylDragon commented on launch/launch/substitutions/python_expression.py:
```diff
@@ -62,11 +66,16 @@ def expression(self) -> List[Substitution]:
+
     def describe(self) -> Text:
         """Return a description of this substitution as a string."""
         return 'PythonExpr({})'.format(' + '.join([sub.describe() for sub in self.expression]))
```
I think the description should be updated to reflect the modules

methylDragon replied:
Likewise, the parse up there, otherwise the XML frontend won't work

See: https://github.com/ros2/launch/blob/rolling/launch/launch/substitutions/launch_configuration.py#L68
In this case the length of data might be 1 or 2. If it's 1, be sure to set 'modules' to the default value in the parse!

RobertBlakeAnderson replied:
I could store the module names in order to print them as part of the Describe output, as printing `__locals` would be way too long for many modules. But then that got me thinking that we might also want to do something like this at Line 53.

> self.__locals = {}
        for x in modules:
            self.__locals[x.__name__] = vars(x)

This way, the expression locals would remain keyed for their respective modules. Then an expression would read `sys.getrefcount(str("hello world!"))` instead of `getrefcount(str("hello world!"))`, for example. This seems good in terms of clarity and avoiding name clashes, especially if someone wants to pass a custom module.

However, doing that would make this PR a breaking change, since PythonExpression currently accepts `math` definitions without the module name. That seems like a reasonable break to make as part of a new ROS2 release, but I wanted to run this idea by you all. I think it would be a good change in the longer run.

RobertBlakeAnderson replied:
Thinking about it even more, the new PythonExpression constructor probably ought to take the module names as a SomeSubstitutionsType, to be consistent with how substitutions generally work.

methylDragon replied:
If we do SomeSubstitutionsType we'll most likely have to take in strings as opposed to module objects directly. I am fine with this since in that case the xml frontend can be unified. I'm not too fussed about this point though.

What will have to change is the parse:
We'll need importing logic, since there isn't a way to pass modules directly in xml.

Pardon the style/naming. I'm just trying to workshop the idea
```
import importlib

    @classmethod
    def parse(cls, data: Iterable[SomeSubstitutionsType]):
        """Parse `PythonExpression` substitution."""
        if not len(data) >= 1 and not len(data) <= 2:
            raise TypeError('eval substitution expects 1 to 2 arguments')
        modules = []
        for i in data[1].split("SOME_DELIMITER_PROBABLY_COMMA"):
            module = importlib.import_module(i)
            modules.append(module)
        # the name 'modules' could change
        return cls, {'expression': data[0], 'modules': module}
```

methylDragon replied:
As for breaking, I think we can have our cake and eat it too. We can just print a warning instead and support both cases.

We can store modules as well as any local names.

More precisely, we can:
```
# Modules is probably not a good name at this point
for x in modules:
    if type(x) is module:
        if any(i for i in vars(x) in in self__locals):
            # PRINT A WARNING
        self.__locals.update(vars(x))
    elif:
        if x.__name__ in self__locals:
            # PRINT A WARNING
        self.__locals[x.__name__] = x
```

methylDragon replied:
Furthermore, if we really really wanted to, we could subdivide self.__locals:

```
self.__locals = {'modules': {}, 'anon_names': {}}
```

And then merge the two dictionaries (and do a warning printout in the case of name clashes) when performing the eval.
(I think this is more efficient than doing the check for each module that is added.

And then when describing the locals dict, you can print the names of any modules, and then all of the anonymous non-module names, to shorten the printout

RobertBlakeAnderson commented:
Got it working with Substitutions in the constructor. Updated tests and added new ones for the XML. I'll now look into the matter of keying the definitions to their modules, with backwards compatibility. As for naming, maybe `python_modules` or `python_deps` would be less ambiguous.

RobertBlakeAnderson commented:
Done. The expressions will now require the module names before the definitions you use. For backwards compatibility, math definitions will work with or without the module name, and the tests include cases for both. If you wish, the implicit references could be considered deprecated for removal at a future date.

I think this PR is ready to go now.

methylDragon commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17848)](http://ci.ros2.org/job/ci_linux/17848/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=12416)](http://ci.ros2.org/job/ci_linux-aarch64/12416/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=18516)](http://ci.ros2.org/job/ci_windows/18516/)

:fire:

methylDragon commented on launch/launch/substitutions/python_expression.py:
```diff
@@ -47,26 +48,61 @@ def __init__(self, expression: SomeSubstitutionsType) -> None:
+            # Check if we got empty list from XML
+            if len(data[1]) > 0:
+                modules_str = data[1][0].perform(None)
+                kwargs['python_modules'] = modules_str.split(', ')
```
Might want to split by "," and then `strip()` each element (so it supports both "a, b, c" and "a,b,c")

methylDragon commented on launch/launch/substitutions/python_expression.py:
```diff
@@ -47,26 +48,61 @@ def __init__(self, expression: SomeSubstitutionsType) -> None:
+            # Check if we got empty list from XML
+            if len(data[1]) > 0:
+                modules_str = data[1][0].perform(None)
+                kwargs['python_modules'] = modules_str.split(', ')
```
```suggestion
                kwargs['python_modules'] = [module.strip() for module in modules_str.split(',')]
```

RobertBlakeAnderson replied:
Good idea. Added that along with a test case.

methylDragon commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17850)](http://ci.ros2.org/job/ci_linux/17850/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=12418)](http://ci.ros2.org/job/ci_linux-aarch64/12418/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=18518)](http://ci.ros2.org/job/ci_windows/18518/)

methylDragon commented:
Flake 8 issues :o

https://ci.ros2.org/job/ci_linux-aarch64/12418/testReport/launch.test/test_flake8/test_flake8/

I think just change the name `locals`

methylDragon commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17854)](http://ci.ros2.org/job/ci_linux/17854/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=12422)](http://ci.ros2.org/job/ci_linux-aarch64/12422/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=18523)](http://ci.ros2.org/job/ci_windows/18523/)
